src/dune_rules/pkg_rules.ml

```diff
@@ -739,9 +826,24 @@ module Action_expander = struct
            ?loc
            [ Pp.text "\"run\" action must have at least one argument" ]
        | (prog_loc, prog) :: args ->
-         let+ exe = Expander.expand_exe_value expander prog ~loc:prog_loc in
-         let args = Value.L.to_strings (List.map ~f:snd args) ~dir in
-         Action.Run (exe, Array.Immutable.of_list args))
+         let+ exe =
+           let prog =
+             match prog with
+             | [ Path prog ] -> Value.Path prog
+             | [ Dir prog ] -> Value.Dir prog
+             | prog ->
+               String (List.map prog ~f:(Value.to_string ~dir) |> String.concat ~sep:"")
```

Comment:
I added a polymorphic variant so that the user knows concatenation is needed. I could add a separate type, but nothing can be done to it apart from converting it back to the list for special handling.

Reply:
Couldn't the new type have a function that does the concatenation given a directory? Are there any other list operations a user may want to do on the values other than this? What's the benefit of exposing the list?

Reply:
We don't always do the concatenation though.

* If we're interpreting a program name, we can special case `[ Path p ]` to avoid it
* When we're interpreting argument names, we preserve the structure of paths.

Now we could add a type like:

```ocaml
module Deferred_value : sig
  type t
  val prog : t -> Path.t
  val args : t -> Run_with_spec.t list
end
```

But I feel like that would artificially expose the implementation details of the expansion in pkg_rules.

Reply:
Ah ok that makes sense then. Thanks for the explanation.